Modules/_io/bytesio.c

```diff
@@ -462,6 +468,38 @@ _io_BytesIO_read1_impl(bytesio *self, Py_ssize_t size)
     return _io_BytesIO_read_impl(self, size);
 }
 
+
+/*[clinic input]
+_io.BytesIO.peek
+    size: Py_ssize_t = 1
+    /
+
+Return bytes from the stream without advancing the position.
+
+If the size argument is zero or negative, read until EOF is reached.
+Return an empty bytes object at EOF.
+[clinic start generated code]*/
+
+static PyObject *
+_io_BytesIO_peek_impl(bytesio *self, Py_ssize_t size)
+/*[clinic end generated code: output=fa4d8ce28b35db9b input=cb06614a3ed0496e]*/
+{
+    Py_ssize_t n;
+
+    CHECK_CLOSED(self);
+
+    /* adjust invalid sizes */
+    n = self->string_size - self->pos;
+    if (size < 1 || size > n) {
+        size = n;
+        if (size < 0)
```

Comment:
Is it possible that ``self->string < self->pos``? It should not happen. I would prefer to replace it with an assertion.

scan_eof():

```
    if (self->pos >= self->string_size)
        return 0;
```

I suppose that it's possible that ``self->pos == self->string_size``, but *pos* should not be greater than *string_size*.

Reply:
I checked this now and indeed `self->string_size < self->pos` can happen after `truncate()` (which shortens the buffer but leaves the position unchanged).

I added a comment.